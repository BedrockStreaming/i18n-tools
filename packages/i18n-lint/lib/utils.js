'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.flatten = undefined;

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _lodash = require('lodash');

var flatten = exports.flatten = function flatten(lang, path) {
  var accumulator = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};

  if ((typeof lang === 'undefined' ? 'undefined' : _typeof(lang)) === 'object') {
    (0, _lodash.forEach)(lang, function (value, key) {
      flatten(value, path ? path + '.' + key : key, accumulator);
    });
  } else {
    accumulator[path] = lang;
  }

  return accumulator;
};