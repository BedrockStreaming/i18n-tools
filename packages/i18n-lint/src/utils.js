import { forEach } from 'lodash';

export const flatten = (lang, path, accumulator = {}) => {
  if (typeof lang === 'object') {
    forEach(lang, (value, key) => {
      flatten(value, path ? `${path}.${key}` : key, accumulator);
    });
  } else {
    accumulator[path] = lang;
  }

  return accumulator;
};
