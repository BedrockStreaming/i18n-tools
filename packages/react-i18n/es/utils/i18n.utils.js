import _get from 'lodash-es/get';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

import { sprintf } from 'sprintf-js';

export var translate = function translate(lang) {
  var i18nNames = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return function (key) {
    var data = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    var number = arguments[2];

    var combineKey = key;
    // Pluralize
    if (typeof number !== 'undefined') {
      combineKey = key + '.' + (number < 2 ? 'one' : 'other');
    }

    var translation = _get(lang, combineKey, combineKey);

    return sprintf(translation, _extends({}, data, i18nNames));
  };
};

export var buildList = function buildList(lang) {
  return function (list, maxSize) {
    var separator = _get(lang, '_i18n.separator');
    var lastSeparator = _get(lang, '_i18n.and');

    if (!list || !list.length) {
      return '';
    }

    if (list.length === 1) {
      return list[0];
    }

    var lastIndex = list.length - 1;
    var subList = list.slice(0, lastIndex).join(separator);
    var result = '' + subList + lastSeparator + list[lastIndex];

    return result.length > maxSize ? result.substring(0, maxSize) + '...' : result;
  };
};