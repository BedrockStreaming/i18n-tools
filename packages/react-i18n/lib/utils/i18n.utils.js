'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.translate = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _sprintfJs = require('sprintf-js');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var translate = exports.translate = function translate(lang) {
  var i18nNames = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return function (key) {
    var data = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    var number = arguments[2];

    var combineKey = key;
    // Pluralize
    if (typeof number !== 'undefined') {
      combineKey = key + '.' + (number < 2 ? 'one' : 'other');
    }

    var translation = _lodash2.default.get(lang, combineKey, combineKey);

    return (0, _sprintfJs.sprintf)(translation, _extends({}, data, i18nNames));
  };
};