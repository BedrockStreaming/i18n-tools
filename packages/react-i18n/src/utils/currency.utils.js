const currencyBuilder = (locale, currency) => new Intl.NumberFormat(locale, { style: 'currency', currency }).format;

export const currencyFunctions = {
  fr: currencyBuilder('fr-FR', 'EUR'),
  en: currencyBuilder('us-EN', 'USD'),
  hu: currencyBuilder('hu-HU', 'HUF'),
  hr: currencyBuilder('hr-HR', 'HRK'),
};
