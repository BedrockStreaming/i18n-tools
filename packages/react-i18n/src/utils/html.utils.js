import React from 'react';
import _ from 'lodash';

const upperCase = /^[A-Z]/;
const tagSearch = /(<.+?>)/;
const tagNameSearch = /<\/?([A-z]+)/;

const parseProps = props => {
  if (!props.length) {
    return null;
  }

  return _.compact(
    props.map(prop => {
      const [key, value = true] = prop.split('=');

      return key ? { key, value: typeof value === 'string' ? JSON.parse(value) : null } : null;
    }),
  );
};

const analyseTag = tag => {
  const isAutoClosing = tag[tag.length - 2] === '/';
  const isClosing = tag[1] === '/';
  const tagName = tagNameSearch.exec(tag)[1];
  const isReactComponent = upperCase.test(tagName);
  const props = isClosing
    ? null
    : parseProps(tag.slice(tag.indexOf(tagName) + tagName.length + 1, tag.length - (isAutoClosing ? 2 : 1)).split(' '));

  return {
    isAutoClosing,
    isClosing,
    isReactComponent,
    tagName,
    props,
    raw: tag,
  };
};

const buildTree = (elements, config = { currentIndex: 0 }, tree = [], currentTagName) => {
  while (config.currentIndex < elements.length) {
    const element = elements[config.currentIndex];

    // eslint-disable-next-line no-param-reassign,no-plusplus
    config.currentIndex++;

    if (typeof element === 'object') {
      if (element.isClosing && element.tagName === currentTagName) {
        return tree;
      }

      if (!element.isAutoClosing && !element.isClosing) {
        element.children = [];
        buildTree(elements, config, element.children, element.tagName);
      }
    }

    tree.push(element);
  }

  if (currentTagName) throw new Error(`Malformated HMTL, can't build proper render`);

  return tree;
};

const buildProps = props => props.reduce((acc, { key, value }) => ({ ...acc, [key]: value }), {});

const renderer = (tree, renderers = {}) =>
  tree.map(node => {
    if (typeof node === 'object') {
      if (node.isReactComponent) {
        if (renderers[node.tagName]) {
          return React.createElement(
            renderers[node.tagName],
            buildProps(node.props),
            node.isAutoClosing ? undefined : renderer(node.children, renderers),
          );
        }

        const fallback = [node.raw];
        if (node.children.length) {
          fallback.push(renderer(node.children, renderers));
          fallback.push(`</${node.tagName}>`);
        }

        return fallback;
      }

      return React.createElement(
        node.tagName,
        buildProps(node.props),
        node.isAutoClosing ? undefined : renderer(node.children, renderers),
      );
    }

    return node;
  });

export const interpolateHTMLTags = (translation, renderers) => {
  const tags = _.compact(translation.split(tagSearch), x => x !== '').reduce(
    (acc, element) => acc.concat(element[0] === '<' ? analyseTag(element) : element),
    [],
  );

  if (tags.length === 1) return tags;

  try {
    const tree = buildTree(tags);

    return renderer(tree, renderers);
  } catch (error) {
    return translation;
  }
};
