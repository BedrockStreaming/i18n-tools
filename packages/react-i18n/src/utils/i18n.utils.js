import _get from 'lodash/get';
import _has from 'lodash/has';
import _noop from 'lodash/noop';
import { sprintf } from 'sprintf-js';
import { interpolateHTMLTags } from './html.utils';

const pluralizeFunctions = {
  de: number => (number === 1 ? 'one' : 'other'),
  en: number => (number === 1 ? 'one' : 'other'),
  fr: number => (number > 1 ? 'other' : 'one'),
  hr: (number, general) => {
    // General plural
    if (general) {
      return number > 1 ? 'other' : 'one';
    }

    const numberInString = number.toString();
    const lastDigit = numberInString.charAt(numberInString.length - 1);

    if ((number > 4 && number < 21) || ['0', '5', '6', '7', '8', '9'].includes(lastDigit)) {
      // Third plural form
      return 'many';
    } else if (lastDigit === '1') {
      // First plural form and singular
      return 'one';
    } else if (lastDigit === '2' || lastDigit === '3' || lastDigit === '4') {
      // Second plural form
      return 'few';
    }

    return '';
  },
  hu: (number, general) => {
    if (!general) {
      return 'one';
    }

    return number > 1 ? 'other' : 'one';
  },
  nl: number => (number === 1 ? 'one' : 'other'),
};

export const translate = (lang, i18nNames = {}, errorCallback = _noop, parseHTML = false) => {
  const pluralize = pluralizeFunctions[_get(lang, '_i18n.lang')] || pluralizeFunctions.fr;

  return (key, { data = {}, number, general, renderers } = {}) => {
    let combineKey = key;
    // Pluralize
    if (typeof number !== 'undefined') {
      combineKey = `${key}.${pluralize(number, general)}`;
    }

    let translation;
    if (_has(lang, combineKey)) {
      translation = _get(lang, combineKey);
    } else {
      errorCallback(combineKey);
      translation = combineKey;
    }

    const translatedResult = sprintf(translation, { ...data, ...i18nNames, number });

    if (!parseHTML) return translatedResult;

    const htmlTags = interpolateHTMLTags(translatedResult, renderers);
    if (htmlTags.length > 1) return htmlTags;

    return translatedResult;
  };
};

export const buildList = lang => (list, maxSize) => {
  const separator = _get(lang, '_i18n.separator');
  const lastSeparator = _get(lang, '_i18n.and');

  if (!list || !list.length) {
    return '';
  }

  if (list.length === 1) {
    return list[0];
  }

  const lastIndex = list.length - 1;
  const subList = list.slice(0, lastIndex).join(separator);
  const result = `${subList}${lastSeparator}${list[lastIndex]}`;

  return result.length > maxSize ? `${result.substring(0, maxSize)}...` : result;
};
