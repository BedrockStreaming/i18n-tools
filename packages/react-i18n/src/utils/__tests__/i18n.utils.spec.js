/* eslint-disable react/prop-types */

import React from 'react';
import { mount } from 'enzyme';
import { translate, buildList } from '../i18n.utils';

describe('i18n translate function', () => {
  it('should return the key if the translation is missing', () => {
    expect(translate()('foo.bar')).toBe('foo.bar');
    expect(translate({})('foo.bar')).toBe('foo.bar');
  });

  it('should return translation for a given key ', () => {
    const lang = { foo: { bar: 'foo bar!' } };
    expect(translate(lang)('foo.bar')).toBe('foo bar!');
  });

  it('should interpolate', () => {
    const lang = { foo: { bar: 'Hello %(name)s !' } };
    const t = translate(lang);
    expect(t('foo.bar', { name: 'World' })).toBe('Hello World !');
  });

  describe('with general plural', () => {
    it('should pluralize with default value', () => {
      const lang = { foo: { bar: { one: 'foo bar!', other: 'foos bars!!!' } } };
      const t = translate(lang);
      expect(t('foo.bar', undefined, 0, true)).toBe('foo bar!');
      expect(t('foo.bar', undefined, 1, true)).toBe('foo bar!');
      expect(t('foo.bar', undefined, 2, true)).toBe('foos bars!!!');
    });

    it('should pluralize in french', () => {
      const lang = {
        foo: { bar: { one: 'foo bar!', other: 'foos bars!!!' } },
        _i18n: { lang: 'fr' },
      };
      const t = translate(lang);
      expect(t('foo.bar', undefined, 0, true)).toBe('foo bar!');
      expect(t('foo.bar', undefined, 1, true)).toBe('foo bar!');
      expect(t('foo.bar', undefined, 2, true)).toBe('foos bars!!!');
    });

    it('should pluralize in english', () => {
      const lang = {
        foo: { bar: { one: 'foo bar!', other: 'foos bars!!!' } },
        _i18n: { lang: 'en' },
      };
      const t = translate(lang);
      expect(t('foo.bar', undefined, 0, true)).toBe('foos bars!!!');
      expect(t('foo.bar', undefined, 1, true)).toBe('foo bar!');
      expect(t('foo.bar', undefined, 2, true)).toBe('foos bars!!!');
    });

    it('should pluralize in hungarian', () => {
      const lang = {
        foo: { bar: { one: 'foo bar!', other: 'foos bars!!!' } },
        _i18n: { lang: 'hu' },
      };
      const t = translate(lang);
      expect(t('foo.bar', undefined, 0, true)).toBe('foo bar!');
      expect(t('foo.bar', undefined, 1, true)).toBe('foo bar!');
      expect(t('foo.bar', undefined, 2, true)).toBe('foos bars!!!');
    });

    it('should pluralize in croatian', () => {
      const lang = {
        foo: {
          bar: {
            other: 'general plural',
            one: 'first plural',
            few: 'second plural',
            many: 'third plural',
          },
        },
        _i18n: { lang: 'hr' },
      };
      const t = translate(lang);

      expect(t('foo.bar', undefined, 0, true)).toBe('first plural');
      expect(t('foo.bar', undefined, 1, true)).toBe('first plural');
      expect(t('foo.bar', undefined, 2, true)).toBe('general plural');
    });
  });

  describe('with number interpolation', () => {
    it('should pluralize with default value', () => {
      const lang = { foo: { bar: { one: '%(number)d foo bar!', other: '%(number)d foos bars!!!' } } };
      const t = translate(lang);
      expect(t('foo.bar', undefined, 0)).toBe('0 foo bar!');
      expect(t('foo.bar', undefined, 1)).toBe('1 foo bar!');
      expect(t('foo.bar', undefined, 2)).toBe('2 foos bars!!!');
    });

    it('should pluralize in french', () => {
      const lang = {
        foo: { bar: { one: '%(number)d foo bar!', other: '%(number)d foos bars!!!' } },
        _i18n: { lang: 'fr' },
      };
      const t = translate(lang);
      expect(t('foo.bar', undefined, 0)).toBe('0 foo bar!');
      expect(t('foo.bar', undefined, 1)).toBe('1 foo bar!');
      expect(t('foo.bar', undefined, 2)).toBe('2 foos bars!!!');
    });

    it('should pluralize in english', () => {
      const lang = {
        foo: { bar: { one: '%(number)d foo bar!', other: '%(number)d foos bars!!!' } },
        _i18n: { lang: 'en' },
      };
      const t = translate(lang);
      expect(t('foo.bar', undefined, 0)).toBe('0 foos bars!!!');
      expect(t('foo.bar', undefined, 1)).toBe('1 foo bar!');
      expect(t('foo.bar', undefined, 2)).toBe('2 foos bars!!!');
    });

    it('should pluralize in english', () => {
      const lang = {
        foo: { bar: { one: '%(number)d foo bar!', other: '%(number)d foos bars!!!' } },
        _i18n: { lang: 'en' },
      };
      const t = translate(lang);
      expect(t('foo.bar', undefined, 0)).toBe('0 foos bars!!!');
      expect(t('foo.bar', undefined, 1)).toBe('1 foo bar!');
      expect(t('foo.bar', undefined, 2)).toBe('2 foos bars!!!');
    });

    it('should pluralize in hungarian', () => {
      const lang = {
        foo: { bar: { one: '%(number)d foo bar!', other: '%(number)d foos bars!!!' } },
        _i18n: { lang: 'hu' },
      };
      const t = translate(lang);
      expect(t('foo.bar', undefined, 0)).toBe('0 foo bar!');
      expect(t('foo.bar', undefined, 1)).toBe('1 foo bar!');
      expect(t('foo.bar', undefined, 2)).toBe('2 foo bar!');
    });

    it('should pluralize in croatian', () => {
      const lang = {
        foo: {
          bar: {
            one: '%(number)d first plural',
            few: '%(number)d second plural',
            many: '%(number)d third plural',
          },
        },
        _i18n: { lang: 'hr' },
      };
      const t = translate(lang);
      // first plural form
      [1, 21, 31, 41, 101].forEach(i => expect(t('foo.bar', undefined, i)).toBe(`${i} first plural`));

      // second plural form
      [2, 3, 4, 22, 23, 24, 32, 33, 34].forEach(i => expect(t('foo.bar', undefined, i)).toBe(`${i} second plural`));

      // third plural form
      for (let i = 5; i < 21; i++) {
        expect(t('foo.bar', undefined, i)).toBe(`${i} third plural`);
      }
      [0, 25, 26, 27, 28, 29, 30, 35, 36, 37, 38, 39, 40, 45].forEach(i =>
        expect(t('foo.bar', undefined, i)).toBe(`${i} third plural`),
      );
    });
  });

  describe('with JSX', () => {
    const Bold = ({ children }) => <strong>{children}</strong>;
    const Italic = ({ children }) => <em>{children}</em>;
    const LineBreak = () => <br />;

    it('should render the JSX component present inside the translation', () => {
      const lang = {
        foo: {
          bar: 'Hello <Bold>Moto</Bold> !',
        },
      };
      const renderers = { Bold };
      const t = translate(lang);

      const result = t('foo.bar', undefined, undefined, false, renderers);
      const wrapper = mount(<div>{result}</div>);

      expect(wrapper).toMatchSnapshot();
    });

    it('should render the nested JSX component present inside the translation', () => {
      const lang = {
        foo: {
          bar: 'Hello <Bold><Italic>Moto</Italic></Bold> !',
        },
      };
      const renderers = { Bold, Italic };
      const t = translate(lang);

      const result = t('foo.bar', undefined, undefined, false, renderers);
      const wrapper = mount(<div>{result}</div>);

      expect(wrapper).toMatchSnapshot();
    });

    it('should render the JSX component and sibling JSX component present inside the translation', () => {
      const lang = {
        foo: {
          bar: '<Italic>Hello</Italic> <Bold>Moto</Bold> !',
        },
      };
      const renderers = { Bold, Italic };
      const t = translate(lang);

      const result = t('foo.bar', undefined, undefined, false, renderers);
      const wrapper = mount(<div>{result}</div>);

      expect(wrapper).toMatchSnapshot();
    });

    it('should render the short (without children) JSX component inside the translation', () => {
      const lang = {
        foo: {
          bar: 'Hello<LineBreak />Moto<LineBreak/> !',
        },
      };
      const renderers = { LineBreak };
      const t = translate(lang);

      const result = t('foo.bar', undefined, undefined, false, renderers);
      const wrapper = mount(<div>{result}</div>);

      expect(wrapper).toMatchSnapshot();
    });

    it('should render a complex JSX structure inside the translation', () => {
      const lang = {
        foo: {
          bar: '<Bold><Italic>Hell</Italic>o</Bold><LineBreak />Moto<LineBreak/> !',
        },
      };
      const renderers = { LineBreak, Bold, Italic };
      const t = translate(lang);

      const result = t('foo.bar', undefined, undefined, false, renderers);
      const wrapper = mount(<div>{result}</div>);

      expect(wrapper).toMatchSnapshot();
    });

    /* eslint-disable no-console */
    it('should throw an error if renderer not provided', () => {
      console.warn = jest.fn();
      const lang = {
        foo: {
          bar: 'Hello <Bold><Italic>Moto</Italic></Bold> !',
        },
      };
      const renderers = { Bold };
      const t = translate(lang);
      t('foo.bar', undefined, undefined, false, renderers);

      expect(console.warn).toHaveBeenCalledWith('No renderer provided for component "Italic"');
    });
    /* eslint-enable no-console */

    it('should not try to interpolate basic html tag', () => {
      const lang = {
        foo: {
          bar: '<em>Hello</em> <strong>Moto</strong> !',
        },
      };
      const renderers = {};
      const t = translate(lang);

      const result = t('foo.bar', undefined, undefined, false, renderers);
      const wrapper = mount(<div>{result}</div>);

      expect(wrapper).toMatchSnapshot();
    });

    it('should works with badly formatted JSX', () => {
      const lang = {
        foo: {
          bar: '<Bold>Toto</Italic>'
        }
      }

      const renderers = {Bold, Italic}
      const t = translate(lang)

      const result = t('foo.bar', undefined, undefined, undefined, renderers)
      const wrapper = mount(<div>{result}</div>)

      expect(wrapper).toMatchSnapshot()
    });
  });
});

describe('i18n listBuilder function', () => {
  const lang = {
    _i18n: {
      separator: ', ',
      and: ' and ',
    },
  };

  const listBuilder = buildList(lang);

  it('should build list with only one element', () => {
    const list = ['foo'];

    expect(listBuilder(list)).toBe('foo');
  });

  it('should build list with two elemetns', () => {
    const list = ['foo', 'bar'];

    expect(listBuilder(list)).toBe('foo and bar');
  });

  it('should build list with more than two elements', () => {
    const list = ['foo', 'bar', 'foobar'];

    expect(listBuilder(list)).toBe('foo, bar and foobar');
  });
});
