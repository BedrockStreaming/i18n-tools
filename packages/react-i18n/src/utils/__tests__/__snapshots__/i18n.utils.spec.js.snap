// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`i18n translate function with JSX should not try to interpolate basic html tag 1`] = `
<div>
  &lt;em&gt;Hello&lt;/em&gt; &lt;strong&gt;Moto&lt;/strong&gt; !
</div>
`;

exports[`i18n translate function with JSX should render a complex JSX structure inside the translation 1`] = `
<div>
  <Bold>
    <strong>
      <Italic>
        <em>
          Hell
        </em>
      </Italic>
      o
    </strong>
  </Bold>
  <LineBreak>
    <br />
  </LineBreak>
  Moto
  <LineBreak>
    <br />
  </LineBreak>
   !
</div>
`;

exports[`i18n translate function with JSX should render the JSX component and sibling JSX component present inside the translation 1`] = `
<div>
  <Italic>
    <em>
      Hello
    </em>
  </Italic>
   
  <Bold>
    <strong>
      Moto
    </strong>
  </Bold>
   !
</div>
`;

exports[`i18n translate function with JSX should render the JSX component present inside the translation 1`] = `
<div>
  Hello 
  <Bold>
    <strong>
      Moto
    </strong>
  </Bold>
   !
</div>
`;

exports[`i18n translate function with JSX should render the nested JSX component present inside the translation 1`] = `
<div>
  Hello 
  <Bold>
    <strong>
      <Italic>
        <em>
          Moto
        </em>
      </Italic>
    </strong>
  </Bold>
   !
</div>
`;

exports[`i18n translate function with JSX should render the short (without children) JSX component inside the translation 1`] = `
<div>
  Hello
  <LineBreak>
    <br />
  </LineBreak>
  Moto
  <LineBreak>
    <br />
  </LineBreak>
   !
</div>
`;

exports[`i18n translate function with JSX should works with badly formatted JSX 1`] = `
<div>
  &lt;Bold&gt;Toto&lt;/Italic&gt;
</div>
`;
